examples/batched-solver-from-files/batched-solver-from-files.cpp

```diff
@@ -0,0 +1,223 @@
+    const auto exec = exec_map.at(executor_string)();  // throws if not valid
+
+    const std::string problem_descr_str = argc >= 3 ? argv[2] : "pores_1";
+    const size_type num_systems = argc >= 4 ? std::atoi(argv[3]) : 2;
+    const size_type num_duplications = argc >= 5 ? std::atoi(argv[4]) : 2;
+    const std::string batch_scaling = argc >= 6 ? argv[5] : "none";
+    // @sect3{Read batch from files}
+    auto data = std::vector<gko::matrix_data<value_type>>(num_systems);
+    std::vector<gko::matrix_data<value_type>> bdata(num_systems);
+    auto scale_data = std::vector<gko::matrix_data<value_type>>(num_systems);
+    for (size_type i = 0; i < data.size(); ++i) {
+        const std::string mat_str = "A.mtx";
+        const std::string fbase =
+            "data/" + problem_descr_str + "/" + std::to_string(i) + "/";
+        std::string fname = fbase + mat_str;
+        std::ifstream mtx_fd(fname);
+        std::string b_str = "b.mtx";
+        data[i] = gko::read_raw<value_type>(mtx_fd);
+        std::string bfname =
+            "data/" + problem_descr_str + "/" + std::to_string(i) + "/" + b_str;
+        std::ifstream b_fd(bfname);
+        bdata[i] = gko::read_raw<value_type>(b_fd);
+        // If necessary, 'scaling vectors' can be provided to diagonal-scale
+        //  a system from the left and the right. For this example, no
+        //  scaling vectors are provided.
+        if (batch_scaling == "explicit") {
+            std::string scale_fname = fbase + "S.mtx";
+            std::ifstream scale_fd(scale_fname);
+            scale_data[i] = gko::read_raw<value_type>(scale_fd);
+        }
+    }
+    auto single_batch = mtx_type::create(exec);
+    single_batch->read(data);
+    // We can duplicate the batch a few times if we wish.
+    std::shared_ptr<mtx_type> A =
+        mtx_type::create(exec, num_duplications, single_batch.get());
+    // Create RHS
+    auto temp_b = vec_type::create(exec);
+    temp_b->read(bdata);
+    auto b = vec_type::create(exec, num_duplications, temp_b.get());
+    // Create initial guess as 0 and copy to device
+    const size_type num_total_systems = num_systems * num_duplications;
+    auto x = vec_type::create(exec);
+    auto host_x = vec_type::create(exec->get_master(), b->get_size());
+    // The number of rows in each system is taken as the 0th dimension of the
+    //  size of the 0th system in the batch.
+    const int num_rows = static_cast<int>(b->get_size().at(0)[0]);
+    for (size_type isys = 0; isys < num_total_systems; isys++) {
+        for (int irow = 0; irow < num_rows; irow++) {
+            host_x->at(isys, irow, 0) = gko::zero<value_type>();
+        }
+    }
+    x->copy_from(host_x.get());
+
+    const real_type reduction_factor{1e-6};
+
+    // @sect3{Create the batch solver factory}
+    // Create a batched solver factory with relevant parameters.
+    auto solver_gen =
+        solver_type::build()
+            .with_max_iterations(500)
```

Comment:
I would also add the `num_shared_vectors` parameter here, and a comment to say what it does.